authz_client_fix: fixed ext_authz http client when request contains content-length greater than 0

gsagula opened the pull request:
*Description*:
This PR fixes https://github.com/envoyproxy/envoy/issues/3865.
*Risk Level*: low
*Testing*: unit and manual testing.
*Docs Changes*: n/a
*Release Notes*: n/a

Signed-off-by: Gabriel <gsagula@gmail.com>

gsagula commented:
`envoy//test/integration:websocket_integration_test  TIMEOUT` seems to be not related to my changes. I will retry it later.

gsagula commented on source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:
```diff
@@ -44,14 +44,15 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
-      headers->addCopy(key, value);
-    } else {
-      headers->addCopy(key, header.second);
+    if (key != Http::Headers::get().ContentLength) {
```
Yes, I have the same concern. With the fix, this is exactly what Datawire has been using so far in Ambassador. I think that the whitelist wouldn't be a bad idea. This way would be up to the operator to decide what to copy over. WDYT?

mattklein123 replied:
Yeah, I would recommend redoing this so it's a whitelist. I can think of a bunch of other places in which this could become a problem.

gsagula replied:
Sounds good! I will do the changes. Thanks for the directions.

gsagula commented:
Oops, CI is not happy. I will look at this issue tomorrow again.

dio commented on source/extensions/filters/http/ext_authz/ext_authz.h:
```diff
@@ -40,25 +40,37 @@ class FilterConfig {
 private:
+  static Http::LowerCaseStrUnorderedSet toRequestHeaders(
+      const Protobuf::RepeatedPtrField<Envoy::ProtobufTypes::String>& request_headers) {
+    Http::LowerCaseStrUnorderedSet allowed_headers(request_headers.begin(), request_headers.end());
```
It seems CI's compiler is not happy with this (and [this line](https://github.com/envoyproxy/envoy/pull/3888/files#diff-2b872196d2450427136eae7022c99d00R41)), should we do transformation first?

I think the below compiles:

```c++
  static Http::LowerCaseStrUnorderedSet toLowerCaseStrings(
      const Protobuf::RepeatedPtrField<Envoy::ProtobufTypes::String>& request_headers) {
    Http::LowerCaseStrUnorderedSet headers;
    std::transform(request_headers.begin(), request_headers.end(),
                   std::inserter(headers, headers.end()),
                   [](const std::string& header) { return Http::LowerCaseString(header); });
    return headers;
  }
```

Hence,

```c++
Http::LowerCaseStrUnorderedSet allowed_headers(toLowerCaseStrings(request_headers));
```
And,

```c++
response_headers_to_remove_(toLowerCaseStrings(request_headers));
```

gsagula replied:
@dio Thanks for looking into this 👍  I thought that transformation would be needed here, but apparently, it is. I will push a commit later!

gsagula commented:
@mattklein123 @dio Just a quick recap. This PR implements the whitelist of headers that are allowed to be sent to the authorization server. I've also updated the documentation and added new tests. Feel free to take a look whenever you get a chance. Thanks!

mattklein123 commented on include/envoy/http/header_map.h:
```diff
@@ -28,12 +31,27 @@ class LowerCaseString {
+struct LowerCaseStringHash {
+  size_t operator()(const LowerCaseString& value) const {
+    return HashUtil::xxHash64(value.get());
+    ;
```
del

mattklein123 commented on include/envoy/http/header_map.h:
```diff
@@ -28,12 +31,27 @@ class LowerCaseString {
 
   std::string string_;
 };
+/**
```
nit: newline before this line

mattklein123 commented on include/envoy/http/header_map.h:
```diff
@@ -28,12 +31,27 @@ class LowerCaseString {
+};
+
+/**
+ * Convinient type for unordered set of lower case string.
```
typo: Convinient

mattklein123 commented on api/envoy/config/filter/http/ext_authz/v2alpha/ext_authz.proto:
```diff
@@ -62,4 +62,10 @@ message HttpService {
   // Sets a list of headers that should be not be sent *from the authorization server* to the
   // upstream.
   repeated string response_headers_to_remove = 3;
```
Shouldn't this also become a whitelist? Aren't we going to have the same problem? IMO it's a lot more clear if we are consistent here and make both a whitelist. Should we switch this? I'm not sure we need to go through the full deprecation cycle for this, but up to you.

gsagula replied:
Let me take a look into this. I remember that there was a bunch of discussions around it, but I agree that whitelist makes much more sense here.

gsagula commented:
@mattklein123 Thanks for the review. I've addressed the issues and also added a configurable list for authorization headers. Please, feel free to take another look.

mattklein123 commented on api/envoy/config/filter/http/ext_authz/v2alpha/ext_authz.proto:
```diff
@@ -42,24 +42,29 @@ message ExtAuthz {
+  // Sets a list of headers that can be sent from the authorization server to the upstream or to the
+  // downstream when present in the authorization response. Note that a matched request header will
+  // have its value overriden by the ones sent from the authorization server.
+  repeated string allowed_authorization_headers = 3;
```
I'm not sure of the production status of this filter? Can we just change this? Or do we need to deprecate and use new fields? (IMO that's probably safer. I don't know that we need to fully deprecate but you could reserve 3 and allocate 4 and 5?)

gsagula commented:
@mattklein123 I think it's good now. One quick question: do you think that integration tests would be reasonable/doable for this filter? Thanks!

mattklein123 commented on api/envoy/config/filter/http/ext_authz/v2alpha/ext_authz.proto:
```diff
@@ -42,24 +42,29 @@ message ExtAuthz {
 
-  // Sets a list of headers that should be not be sent *from the authorization server* to the
-  // upstream.
-  repeated string response_headers_to_remove = 3;
```
Please reserve field 3. `reserve 3;` IIRC. See other examples in the protos.

gsagula replied:
Sorry, I misunderstood your previous instructions. Pushing the changes shortly.

mattklein123 commented on api/envoy/config/filter/http/ext_authz/v2alpha/ext_authz.proto:
```diff
@@ -42,24 +42,31 @@ message ExtAuthz {
-// headers. The HTTP service also allows the authorization filter to also pass data from the
-// response body to the downstream client in case of a denied request.
+// unless a different status code has been indicated by the authorization server via response
+// headers. If other headers in the authorization response need to be send to client, this can also
```
s/send/sent

gsagula replied:
Damn typos, sorry. I will do a pass on everything again.

gsagula commented:
@mattklein123 Should be good now 🤞Thanks!
